### PR TITLE
Fix #30: Decouple VideoProcessor/Monitor from Workers

### LINKED ISSUE #30: Decouple VideoProcessor from Workers
I'd like us to consider having a self contained `VideoProcessor/Monitor` object that's not tied to a Worker object.

The idea is to have a processing model similar to [`OffscrenCanvas`](https://wiki.whatwg.org/wiki/OffscreenCanvas), where you have an object you can use either on the main thread, or in a worker context.

Events would be dispatched on this `VideoProcessor/Monitor` object instead of extending the `Worker` interface. The `VideoProcessor/Monitor` would be `Transferable` to Workers using `postMessage`.
Those objects would be created using a `createVideoProcessor/Monitor` method on `MediaStreamTrack`s. You can release them using a `close` method.

We'd need to figure out a way to get the output `MediaStreamTrack` of a `VideoProcessor`, but that's a little detail.

### PR DESCRIPTION
**Pull request by anssiko**
A quick rewrite based on the suggestions in #30. The idea is to test how this approach resonates with people, solicit further feedback, and if there's agreement polish further. Beware, has rough edges.

[rawgit.com HTML preview with changes](https://rawgit.com/w3c/mediacapture-worker/issue-30-decouple-from-workers/index.html)

### CODE REVIEW

**Comment by anssiko** [comment]
@ChiahungTai @kakukogou PTAL.

If this looks good, we could land this first and then you could rebase #53 and #54.

File: index.html [hunk 1]
```diff
@@ -352,132 +348,145 @@
...
+      </section>
+    </section>
+    <section>
+      <h2>
+        <code>VideoMonitorEvent</code> interface
+      </h2>
+      <p>
+        The <a>video monitor event</a> contains an <a>input frame</a> and its
+        metadata originating from the <a>input media stream track</a>. It uses
+        the <a>VideoMonitorEvent</a> interface for its <a>videomonitor</a>
```
Comment (kakukogou) [inline index.html]: Should here be 

```
...... for its <a>videomonitorchange</a>
```

?

File: index.html [hunk 2]
```diff
@@ -489,49 +498,47 @@
...
+        The <a>video processor event</a> inherits from the <a>video monitor
+        event</a>, and in addition provides means to programmatically construct
+        an <a>output frame</a> for the <a>output media stream track</a>. It
+        uses the <a>VideoProcessorEvent</a> interface for its
+        <a>videoprocessorchange</a> events:
+      </p>
+      <pre class="idl">
+          [Constructor(DOMString type, optional VideoProcessorEventInit videoProcessorEventInitDict)]
           interface VideoProcessorEvent : VideoMonitorEvent {
                           attribute ImageBitmap? outputImageBitmap;
```
Comment (kakukogou) [inline index.html:542]: Just make a note that the `outputImageBitmap` attribute should have type `Promise<ImageBitmap?>`, see #38. Or, even replace this attribute with a setting method, see #39. Of course, this issue does not belong to the scope of this PR, we should handle it separately.

File: index.html [hunk 3]
```diff
@@ -540,45 +547,44 @@
...
-            Set <var>output bitmap</var> as the <a>source</a> of the
-            <var>output media stream track</var>.
-          </li>
-        </ol>
+      <div>
         <p>
-          When a user agent is required to <dfn>fire a video processor
-          event</dfn>, it must <a>fire a video worker event</a> named
-          <a>videoprocess</a>.
+          The <dfn>outputImageBitmap</dfn> attribute must return the value it
```
Comment (kakukogou) [inline index.html]: `trackID`, `playbackTime` and `inputImageBitmap` use `<dfn><code>...</code></dfn>`. Should `outputImageBitmap` use the same style?

File: index.html [hunk 4]
```diff
@@ -279,71 +282,64 @@
...
+        the <a>outputImageBitmap</a> attribute of the
+        <a>VideoProcessorEvent</a> dispatched at <a>video processor</a>.
       </p>
       <p>
-        The <dfn>input media stream track</dfn> for a <a>video worker</a>
-        <var>w</var> is a <a>MediaStreamTrack</a> object passed as the first
-        argument to the <code><a>addVideoMonitor</a>()</code> or
-        <code><a>addVideoProcessor</a>()</code> method on <var>w</var>.
+        The <dfn>input media stream track</dfn> is a <a>MediaStreamTrack</a>
+        object passed as the first argument to the
```
Comment (kakukogou) [inline index.html]: I am confused with this statement, the first argument of both `addVideoMonitor()` and `addVideoProcessor()` is not a `MediaStreamTrack`, so how could the `input media stream track` be defined in this way?

Should it be something likes this?

```
The <dfn>input media stream track</dfn> is the <a>MediaStreamTrack</a> object on which the <code><a>addVideoMonitor</a>()</code> or <code><a>addVideoProcessor</a>()</code> method is invoked.
```

**Comment by kakukogou** [comment]
LGTM with the comments. Let's wait for @ChiahungTai 's review, he might have some feedbacks from implementation.

**Comment by anssiko** [comment]
@kakukogou Thanks for the comments! I'll fix them all at once after getting @ChiahungTai feedback as well.

**Comment by ChiahungTai** [comment]
I hope I can review this patch next week. Busy at fixing regression and test case fail in my re-factoring bug.  :(

File: index.html [hunk 5]
```diff
@@ -352,132 +348,145 @@
...
+        The <a>VideoMonitor</a> interface is used to analyze and
+        <a>VideoProcessor</a> interface in addition to generate and process
+        video data directly using JavaScript. The <a data-lt=
+        "video monitor event">video monitor events</a> are dispatched at
+        <a>VideoMonitor</a> objects, and <a data-lt=
+        "video processor event">video processor events</a> are dispatched at
+        <a>VideoProcessor</a> objects.
+      </p>
+      <pre class="idl">
+          [Constructor]
```
Comment (kakukogou) [inline index.html]: The constructor should accept a script url and should expose to both Window and Worker, right?

File: index.html (hunk 4 above)
Comment (ChiahungTai) [inline index.html]: Kaku is right.

File: index.html (hunk 2 above)
Comment (ChiahungTai) [inline index.html:542]: Yes, we can change to Promise based later.

File: index.html [hunk 6]
```diff
@@ -1621,132 +1577,100 @@
       <h2>
         Examples
       </h2>
```
Comment (ChiahungTai) [inline index.html:1579]: @anssiko 
Should we add a non-worker example right here?

**Comment by anssiko** [comment]
@ChiahungTai @kakukogou thanks for the review!

**Pull request merged**